pkg/agent/flowexporter/exporter/exporter.go

```diff
@@ -147,57 +145,82 @@ func NewFlowExporter(connStore *connections.ConntrackConnectionStore, records *f
-// based on active flow and idle flow timeouts.
 func (exp *flowExporter) Run(stopCh <-chan struct{}) {
-	go wait.Until(exp.Export, time.Second, stopCh)
-
-	<-stopCh
-}
-
-func (exp *flowExporter) Export() {
-	// Retry to connect to IPFIX collector if the exporting process gets reset
-	if exp.process == nil {
-		err := exp.initFlowExporter()
-		if err != nil {
-			klog.Errorf("Error when initializing flow exporter: %v", err)
-			// There could be other errors while initializing flow exporter other than connecting to IPFIX collector,
-			// therefore closing the connection and resetting the process.
+	defaultTimeout := exp.conntrackPriorityQueue.ActiveFlowTimeout
+	expireTimer := time.NewTimer(defaultTimeout)
+	for {
+		select {
+		case <-stopCh:
 			if exp.process != nil {
 				exp.process.CloseConnToCollector()
-				exp.process = nil
 			}
+			expireTimer.Stop()
 			return
+		case <-expireTimer.C:
+			if exp.process == nil {
+				err := exp.initFlowExporter()
+				if err != nil {
+					klog.ErrorS(err, "Error when initializing flow exporter")
+					// There could be other errors while initializing flow exporter
+					// other than connecting to IPFIX collector, therefore closing
+					// the connection and resetting the process.
+					if exp.process != nil {
+						exp.process.CloseConnToCollector()
+						exp.process = nil
+					}
+					// Initializing flow exporter fails, will retry in next cycle.
+					expireTimer.Reset(defaultTimeout)
+					continue
+				}
+			}
+			// Pop out the expired connections from the conntrack priority queue
+			// and the deny priority queue, and send the data records.
+			nextExpireTime, err := exp.sendFlowRecords()
+			if err != nil {
+				klog.ErrorS(err, "Error when sending expired flow records")
+				// If there is an error when sending flow records because of intermittent
+				// connectivity, we reset the connection to IPFIX collector and retry
+				// in the next export cycle to reinitialize the connection and send flow records.
+				exp.process.CloseConnToCollector()
+				exp.process = nil
+				expireTimer.Reset(defaultTimeout)
+				continue
+			}
+			expireTimer.Reset(nextExpireTime)
 		}
 	}
-	// Send flow records to IPFIX collector.
-	err := exp.sendFlowRecords()
-	if err != nil {
-		klog.Errorf("Error when sending flow records: %v", err)
-		// If there is an error when sending flow records because of intermittent connectivity, we reset the connection
-		// to IPFIX collector and retry in the next export cycle to reinitialize the connection and send flow records.
-		exp.process.CloseConnToCollector()
-		exp.process = nil
-		return
+}
+
+func (exp *flowExporter) sendFlowRecords() (time.Duration, error) {
+	currTime := time.Now()
+	expiredConns1, expireTime1 := exp.conntrackConnStore.GetExpiredConns(currTime)
+	expiredConns2, expireTime2 := exp.denyConnStore.GetExpiredConns(currTime)
+	expiredConns := append(expiredConns1, expiredConns2...)
```

Comment:
this will create unnecessary copies, maybe `GetExpiredConns` could take the slice as a parameter to avoid that.
```go
expiredConns := make([]flowexporter.Connection, 0)
var expireTime1, expireTime1 time.Time
expiredConns, expireTime1 = exp.conntrackConnStore.GetExpiredConns(expiredConns, currTime)
expiredConns, expireTime2 = exp.denyConnStore.GetExpiredConns(expiredConns, currTime)
```
maybe worth checking that it improves performance of `sendFlowRecords` for a large number of connections

Reply:
Make sense to me. Done.